File: src/org/mtransit/parser/mt/data/MFrequency.java
package org.mtransit.parser.mt.data;

import org.mtransit.parser.CleanUtils;
import org.mtransit.parser.Constants;

public class MFrequency implements Comparable<MFrequency> {

	private String serviceId;
	private long tripId;
	private int startTime;
	private int endTime;
	public int headwayInSec;

	public MFrequency(String serviceId, long routeId, long tripId, int startTime, int endTime, int headwayInSec) {
		this.serviceId = serviceId;
		this.tripId = tripId;
		this.startTime = startTime;
		this.endTime = endTime;
		this.headwayInSec = headwayInSec;
		this.uuid = null;
	}

	@Override
	public String toString() {
		StringBuilder sb = new StringBuilder(); //
		sb.append(Constants.STRING_DELIMITER).append(CleanUtils.escape(this.serviceId)).append(Constants.STRING_DELIMITER); // service ID
		sb.append(Constants.COLUMN_SEPARATOR); //
		sb.append(this.tripId); // trip ID
		sb.append(Constants.COLUMN_SEPARATOR); //
		sb.append(this.startTime); // start time
		sb.append(Constants.COLUMN_SEPARATOR); //
		sb.append(this.endTime); // end time
		sb.append(Constants.COLUMN_SEPARATOR); //
		sb.append(this.headwayInSec); // headway in seconds
		return sb.toString();
	}

	private String uuid = null;

	public String getUID() {
		if (this.uuid == null) {
			this.uuid = this.serviceId + Constants.UUID_SEPARATOR + this.tripId + Constants.UUID_SEPARATOR + this.startTime + Constants.UUID_SEPARATOR
					+ this.endTime;
		}
		return this.uuid;
	}

	@Override
	public int compareTo(MFrequency otherFrequency) {
		if (!this.serviceId.equals(otherFrequency.serviceId)) {
			return this.serviceId.compareTo(otherFrequency.serviceId);
		}
		if (this.tripId != otherFrequency.tripId) {
			return Long.compare(this.tripId, otherFrequency.tripId);
		}
		if (this.startTime != otherFrequency.startTime) {
			return this.startTime - otherFrequency.startTime;
		}
		if (this.endTime != otherFrequency.endTime) {
			return this.endTime - otherFrequency.endTime;
		}
		return this.headwayInSec - otherFrequency.headwayInSec;
	}

	@Override
	public boolean equals(Object obj) {
		MFrequency ts = (MFrequency) obj;
		if (ts.serviceId != null && !ts.serviceId.equals(serviceId)) {
			return false;
		}
		// no route ID, just for file split
		if (ts.tripId != 0 && ts.tripId != tripId) {
			return false;
		}
		if (ts.startTime != 0 && ts.startTime != startTime) {
			return false;
		}
		if (ts.endTime != 0 && ts.endTime != endTime) {
			return false;
		}
		if (ts.headwayInSec != 0 && ts.headwayInSec != headwayInSec) {
			return false;
		}
		return true;
	}

}

File: src/main/java/org/mtransit/parser/mt/data/MFrequency.kt
package org.mtransit.parser.mt.data

import androidx.annotation.Discouraged
import org.mtransit.commons.sql.SQLUtils
import org.mtransit.parser.gtfs.GAgencyTools
import org.mtransit.parser.gtfs.data.GIDs

data class MFrequency(
    val serviceIdInt: Int,
    private val directionId: Long, // exported
    val startTime: Int,
    val endTime: Int,
    private val headwayInSec: Int
) : Comparable<MFrequency?> {

    @Suppress("unused")
    @get:Discouraged(message = "Not memory efficient")
    val serviceId: String get() = _serviceId

    private val _serviceId: String
        get() = GIDs.getString(serviceIdInt)

    val uID by lazy { getNewUID(serviceIdInt, directionId, startTime, endTime) }

    fun toFile(agencyTools: GAgencyTools) = buildList {
        add(MServiceIds.convert(agencyTools.cleanServiceId(_serviceId)))
        add(directionId.toString())
        add(startTime.toString())
        add(endTime.toString())
        add(headwayInSec.toString())
    }.joinToString(SQLUtils.COLUMN_SEPARATOR)

    override fun compareTo(other: MFrequency?): Int {
        return when {
            other !is MFrequency -> +1
            serviceIdInt != other.serviceIdInt -> _serviceId.compareTo(other._serviceId)
            directionId != other.directionId -> directionId.compareTo(other.directionId)
            startTime != other.startTime -> startTime - other.startTime
            endTime != other.endTime -> endTime - other.endTime
            else -> headwayInSec - other.headwayInSec
        }
    }

    fun toStringPlus(): String {
        return toString() +
                "+(serviceId:$_serviceId)" +
                "+(uID:$uID)"
    }

    companion object {

        private const val UID_SEPARATOR = "+" // int IDs can be negative

        @JvmStatic
        fun getNewUID(
            serviceIdInt: Int,
            directionId: Long,
            startTime: Int,
            endTime: Int
        ) = "${serviceIdInt}$UID_SEPARATOR${directionId}$UID_SEPARATOR${startTime}$UID_SEPARATOR${endTime}"
    }
}